BT Career Challenge

1. Check the data, adjust data, delete null value

1.1 Import needed libraries and data

In [1]:
# import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn

# functions for modelling
from sklearn.preprocessing import OneHotEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score

d:\Software\Anaconda\InstallEvi\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.3
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [2]:
# use pandas import data
df_customer = pd.read_csv('./Data/Data Set.csv')
df_customer

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     3668-QPYBK    Male              0      No         No       2   
2     9237-HQITU  Female              0      No         No       2   
3     9305-CDSKC  Female              0      No         No       8   
4     1452-KIOVK    Male              0      No        Yes      22   
...          ...     ...            ...     ...        ...     ...   
7038  4807-IZYOZ  Female              0      No         No      51   
7039  9710-NJERN  Female              0      No         No      39   
7040  9281-CEDRU  Female              0     Yes         No      68   
7041  2569-WGERO  Female              0      No         No      72   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService       OnlineSecurity  ...  \
0              No  No phone service             DSL                   No  ...   
1             Yes                No             DSL                  Yes  ...   
2             Yes                No     Fiber optic                   No  ...   
3             Yes               Yes     Fiber optic                   No  ...   
4             Yes               Yes     Fiber optic                   No  ...   
...           ...               ...             ...                  ...  ...   
7038          Yes                No              No  No internet service  ...   
7039          Yes                No              No  No internet service  ...   
7040          Yes                No             DSL                   No  ...   
7041          Yes                No              No  No internet service  ...   
7042          Yes                No     Fiber optic                  Yes  ...   

         DeviceProtection          TechSupport          StreamingTV  \
0                      No                   No                   No   
1                      No                   No                   No   
2                      No                   No                   No   
3                     Yes                   No                  Yes   
4                      No                   No                  Yes   
...                   ...                  ...                  ...   
7038  No internet service  No internet service  No internet service   
7039  No internet service  No internet service  No internet service   
7040                   No                  Yes                  Yes   
7041  No internet service  No internet service  No internet service   
7042                  Yes                  Yes                  Yes   

          StreamingMovies      Contract PaperlessBilling  \
0                      No  Monthtomonth              Yes   
1                      No  Monthtomonth              Yes   
2                      No  Monthtomonth              Yes   
3                     Yes  Monthtomonth              Yes   
4                      No  Monthtomonth              Yes   
...                   ...           ...              ...   
7038  No internet service      Two year               No   
7039  No internet service      Two year               No   
7040                   No      Two year               No   
7041  No internet service      Two year              Yes   
7042                  Yes      Two year              Yes   

                  PaymentMethod MonthlyCharges  TotalCharges Churn  
0              Electronic check          29.85         29.85    No  
1                  Mailed check          53.85        108.15   Yes  
2              Electronic check          70.70        151.65   Yes  
3              Electronic check          99.65         820.5   Yes  
4       Credit card (automatic)          89.10        1949.4    No  
...                         ...            ...           ...   ...  
7038  Bank transfer (automatic)          20.65       1020.75    No  
7039               Mailed check          20.15           826    No

In [3]:
# check the datatype and the number of records
df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [4]:
# check the number of churn
df_customer.Churn.value_counts()

No     5174
Yes    1869
Name: Churn, dtype: int64

From the inital exploration, we can see we have 21 columns covering customers' features, and the most important feature churn has 1869 customers who left BT services (indicating it is an imbalanced dataset). It is time to validate data including:
- delete unuseful columns, like customerID
- delete null values
- check if customers is unique
- convert some features from string to numeric values
- merge some categories in features

1.2 Sort out data

To process logistic regression, the values in columns should be transformed:
- Change data type
- Merge categories
- Generate dummy variable for categorical variables (One hot encoding)

In [5]:
# drop customerID
df_customer.drop('customerID', axis = 'columns', inplace = True)
df_customer.sample(5)

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
1064  Female              1     Yes        Yes      22          Yes   
540   Female              0      No         No      20          Yes   
1425  Female              0      No         No       1          Yes   
6488  Female              0      No         No      56          Yes   
3412    Male              0     Yes         No       5          Yes   

     MultipleLines InternetService       OnlineSecurity         OnlineBackup  \
1064           Yes     Fiber optic                   No                   No   
540             No              No  No internet service  No internet service   
1425            No     Fiber optic                   No                   No   
6488            No              No  No internet service  No internet service   
3412            No              No  No internet service  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
1064                   No                   No                  Yes   
540   No internet service  No internet service  No internet service   
1425                   No                   No                   No   
6488  No internet service  No internet service  No internet service   
3412  No internet service  No internet service  No internet service   

          StreamingMovies      Contract PaperlessBilling  \
1064                   No  Monthtomonth               No   
540   No internet service  Monthtomonth              Yes   
1425                   No  Monthtomonth               No   
6488  No internet service      Two year               No   
3412  No internet service  Monthtomonth               No   

                  PaymentMethod  MonthlyCharges TotalCharges Churn  
1064  Bank transfer (automatic)           85.35       1961.6    No  
540   Bank transfer (automatic)           19.50          413    No  
1425           Electronic check           70.10         70.1    No  
6488  Bank transfer (automatic)           19.70       1051.9    No  
3412               Mailed check           19.35       126.05   Yes

In [6]:
# convert TotalCharges from categorical values to numerical values
df_customer["TotalCharges"] = pd.to_numeric(df_customer.TotalCharges,errors = 'coerce')

In [7]:
# check the unique values of each feature
def print_unique_col_values(df):
    for column in df:
        if df[column].dtype == 'object':
            print(f'{column}: {df[column].unique()}')

print_unique_col_values(df_customer)

gender: ['Female' 'Male']
Partner: ['Yes' 'No']
Dependents: ['No' 'Yes']
PhoneService: ['No' 'Yes']
MultipleLines: ['No phone service' 'No' 'Yes']
InternetService: ['DSL' 'Fiber optic' 'No']
OnlineSecurity: ['No' 'Yes' 'No internet service']
OnlineBackup: ['Yes' 'No' 'No internet service']
DeviceProtection: ['No' 'Yes' 'No internet service']
TechSupport: ['No' 'Yes' 'No internet service']
StreamingTV: ['No' 'Yes' 'No internet service']
StreamingMovies: ['No' 'Yes' 'No internet service']
Contract: ['Monthtomonth' 'Two year' 'One year']
PaperlessBilling: ['Yes' 'No']
PaymentMethod: ['Electronic check' 'Mailed check' 'Credit card (automatic)'
 'Bank transfer (automatic)']
Churn: ['No' 'Yes']


In [8]:
# replace No internet service to No, replace No phone service to No
df_customer.replace('No phone service', 'No', inplace = True)
df_customer.replace('No internet service', 'No', inplace = True)
df_customer.sample(5)

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
2162    Male              0      No         No      19          Yes   
3371  Female              0     Yes         No      25          Yes   
3760    Male              1      No         No       6          Yes   
6676    Male              0     Yes         No      70          Yes   
6202    Male              0     Yes        Yes      71          Yes   

     MultipleLines InternetService OnlineSecurity OnlineBackup  \
2162            No     Fiber optic             No           No   
3371           Yes             DSL             No          Yes   
3760            No     Fiber optic             No           No   
6676           Yes              No             No           No   
6202            No     Fiber optic            Yes          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies      Contract  \
2162              Yes         Yes         Yes              No  Monthtomonth   
3371               No          No          No              No  Monthtomonth   
3760               No          No         Yes              No  Monthtomonth   
6676               No          No          No              No      Two year   
6202              Yes         Yes         Yes              No      Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges  \
2162               No               Mailed check           89.95   
3371               No           Electronic check           54.10   
3760              Yes           Electronic check           79.70   
6676               No  Bank transfer (automatic)           24.25   
6202               No    Credit card (automatic)          100.20   

      TotalCharges Churn  
2162       1682.40    No  
3371       1373.00    No  
3760        497.60    No  
6676       1724.15    No  
6202       7209.00    No

In [9]:
# convert yes to 1, no to 0 for analysis; male to 1, female to 0 for analysis
mapping_dict = {'Male': 0, 'Female': 1, 'No': 0, 'Yes': 1}

cols_to_replace = ['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 
                   'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 
                   'TechSupport', 'StreamingTV', 'StreamingMovies', 'PaperlessBilling', 'Churn']

df_customer[cols_to_replace] = df_customer[cols_to_replace].replace(mapping_dict)
df_customer.sample(5)


gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
379        1              0        0           0       1             1   
5066       0              0        1           0      40             1   
4519       1              0        0           0      39             1   
6263       0              1        1           0      64             1   
5018       1              0        0           0      26             1   

      MultipleLines InternetService  OnlineSecurity  OnlineBackup  \
379               0             DSL               0             0   
5066              0     Fiber optic               0             1   
4519              0             DSL               0             1   
6263              1              No               0             0   
5018              0             DSL               0             1   

      DeviceProtection  TechSupport  StreamingTV  StreamingMovies  \
379                  0            1            0                0   
5066                 1            0            0                0   
4519                 0            0            1                0   
6263                 0            0            0                0   
5018                 1            0            0                0   

          Contract  PaperlessBilling            PaymentMethod  MonthlyCharges  \
379   Monthtomonth                 1         Electronic check           49.90   
5066      One year                 0         Electronic check           80.80   
4519      One year                 1  Credit card (automatic)           58.60   
6263      Two year                 0  Credit card (automatic)           25.65   
5018      One year                 1             Mailed check           56.05   

      TotalCharges  Churn  
379          49.90      0  
5066       3132.75      0  
4519       2224.50      0  
6263       1740.80      0  
5018       1553.20      0

In [10]:
# delete records with missing values
df_customer.dropna()
df_customer.sample(5)

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
3725       0              0        1           0       3             0   
6676       0              0        1           0      70             1   
3777       1              0        0           0      37             1   
6011       0              0        1           1      42             1   
2279       1              0        0           0      40             1   

      MultipleLines InternetService  OnlineSecurity  OnlineBackup  \
3725              0             DSL               0             0   
6676              1              No               0             0   
3777              1     Fiber optic               0             1   
6011              0              No               0             0   
2279              0     Fiber optic               0             1   

      DeviceProtection  TechSupport  StreamingTV  StreamingMovies  \
3725                 0            1            0                0   
6676                 0            0            0                0   
3777                 1            0            1                1   
6011                 0            0            0                0   
2279                 0            0            0                0   

          Contract  PaperlessBilling              PaymentMethod  \
3725  Monthtomonth                 1           Electronic check   
6676      Two year                 0  Bank transfer (automatic)   
3777  Monthtomonth                 1  Bank transfer (automatic)   
6011      Two year                 0               Mailed check   
2279  Monthtomonth                 0    Credit card (automatic)   

      MonthlyCharges  TotalCharges  Churn  
3725           29.90         92.25      0  
6676           24.25       1724.15      0  
3777          101.90       3545.35      1  
6011           20.70        828.85      0  
2279           74.80       2971.70      0

In [11]:
# generate dummy variables for categorical variables
df_customer_clean = pd.get_dummies(df_customer,columns=['InternetService','Contract','PaymentMethod'])
df_customer_clean.sample(5)

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
5793       0              1        1           0      72             1   
3608       0              0        0           0      24             1   
954        1              0        0           0       9             1   
4463       0              0        0           1       7             1   
4995       0              0        0           1      25             0   

      MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  ...  \
5793              1               0             1                 0  ...   
3608              0               1             0                 0  ...   
954               0               1             0                 0  ...   
4463              0               0             0                 0  ...   
4995              0               1             0                 0  ...   

      InternetService_DSL  InternetService_Fiber optic  InternetService_No  \
5793                    0                            1                   0   
3608                    1                            0                   0   
954                     0                            1                   0   
4463                    0                            0                   1   
4995                    1                            0                   0   

      Contract_Monthtomonth  Contract_One year  Contract_Two year  \
5793                      0                  0                  1   
3608                      1                  0                  0   
954                       1                  0                  0   
4463                      0                  1                  0   
4995                      0                  1                  0   

      PaymentMethod_Bank transfer (automatic)  \
5793                                        1   
3608                                        0   
954                                         0   
4463                                        0   
4995                                        0   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
5793                                      0                               0   
3608                                      0                               1   
954                                       1                               0   
4463                                      0                               0   
4995                                      1                               0   

      PaymentMethod_Mailed check  
5793                           0  
3608                           0  
954                            0  
4463                           1  
4995                           0  

[5 rows x 27 columns]

In [12]:
# scaling data to avoid bias
column_to_scale = ['tenure','MonthlyCharges', 'TotalCharges']

from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
df_customer_clean[column_to_scale] = scaler.fit_transform(df_customer_clean[column_to_scale])

In [24]:
# final check
df_customer_clean.sample(5)

gender  SeniorCitizen  Partner  Dependents    tenure  PhoneService  \
744        1              0        0           0  0.125000             1   
3878       1              0        0           0  0.930556             1   
3979       1              0        1           0  1.000000             1   
6725       1              0        1           1  0.666667             0   
246        1              1        0           0  0.680556             1   

      MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  ...  \
744               0               0             0                 0  ...   
3878              1               1             1                 1  ...   
3979              1               0             1                 1  ...   
6725              0               1             1                 0  ...   
246               1               1             0                 0  ...   

      InternetService_DSL  InternetService_Fiber optic  InternetService_No  \
744                     0                            0                   1   
3878                    0                            1                   0   
3979                    0                            1                   0   
6725                    1                            0                   0   
246                     0                            1                   0   

      Contract_Monthtomonth  Contract_One year  Contract_Two year  \
744                       1                  0                  0   
3878                      1                  0                  0   
3979                      0                  1                  0   
6725                      0                  0                  1   
246                       1                  0                  0   

      PaymentMethod_Bank transfer (automatic)  \
744                                         0   
3878                                        0   
3979                                        1   
6725                                        0   
246                                         0   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
744                                       0                               1   
3878                                      1                               0   
3979                                      0                               0   
6725                                      0                               0   
246                                       1                               0   

      PaymentMethod_Mailed check  
744                            0  
3878                           0  
3979                           0  
6725                           1  
246                            0  

[5 rows x 27 columns]

2. Modelling

After preparing data, there are several steps to build model:
- Split data into training dataset and test dataset

2.1 Split data

In [14]:
# keep columns except churn, copy it to avoid modifying the original one, set datatype as float
X = df_customer_clean.drop(columns = ['Churn']).copy().astype(float)


y = df_customer_clean.Churn.astype(np.float32)

In [15]:
X.sample(5)

gender  SeniorCitizen  Partner  Dependents    tenure  PhoneService  \
6667     1.0            1.0      0.0         0.0  0.930556           1.0   
6899     1.0            0.0      1.0         0.0  1.000000           0.0   
2979     0.0            0.0      0.0         0.0  0.027778           1.0   
1685     1.0            0.0      0.0         0.0  0.138889           1.0   
4988     1.0            0.0      1.0         1.0  0.805556           1.0   

      MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  ...  \
6667            0.0             1.0           1.0               0.0  ...   
6899            0.0             0.0           1.0               1.0  ...   
2979            0.0             0.0           0.0               0.0  ...   
1685            0.0             0.0           0.0               0.0  ...   
4988            1.0             0.0           1.0               1.0  ...   

      InternetService_DSL  InternetService_Fiber optic  InternetService_No  \
6667                  0.0                          1.0                 0.0   
6899                  1.0                          0.0                 0.0   
2979                  0.0                          1.0                 0.0   
1685                  0.0                          0.0                 1.0   
4988                  0.0                          1.0                 0.0   

      Contract_Monthtomonth  Contract_One year  Contract_Two year  \
6667                    0.0                0.0                1.0   
6899                    0.0                0.0                1.0   
2979                    1.0                0.0                0.0   
1685                    1.0                0.0                0.0   
4988                    0.0                1.0                0.0   

      PaymentMethod_Bank transfer (automatic)  \
6667                                      1.0   
6899                                      0.0   
2979                                      0.0   
1685                                      0.0   
4988                                      0.0   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
6667                                    0.0                             0.0   
6899                                    0.0                             1.0   
2979                                    0.0                             1.0   
1685                                    0.0                             0.0   
4988                                    1.0                             0.0   

      PaymentMethod_Mailed check  
6667                         0.0  
6899                         0.0  
2979                         0.0  
1685                         1.0  
4988                         0.0  

[5 rows x 26 columns]

In [25]:
y.value_counts()

0.0    5174
1.0    1869
Name: Churn, dtype: int64

In [26]:
# split training and test sets
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(X, y, test_size=0.20, random_state=23, stratify = y)

In [27]:
y_train.value_counts()

0.0    4139
1.0    1495
Name: Churn, dtype: int64

In [28]:
y_test.value_counts()

0.0    1035
1.0     374
Name: Churn, dtype: int64

In [29]:
X_train.shape

(5634, 26)

In [30]:
X_train.sample(5)

gender  SeniorCitizen  Partner  Dependents    tenure  PhoneService  \
6290     1.0            0.0      1.0         0.0  1.000000           1.0   
520      1.0            0.0      0.0         0.0  0.111111           1.0   
5748     1.0            0.0      1.0         1.0  0.319444           1.0   
2315     0.0            1.0      0.0         0.0  0.166667           1.0   
1412     1.0            0.0      0.0         0.0  0.013889           1.0   

      MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  ...  \
6290            0.0             0.0           0.0               0.0  ...   
520             0.0             0.0           1.0               0.0  ...   
5748            1.0             0.0           0.0               0.0  ...   
2315            1.0             0.0           0.0               0.0  ...   
1412            1.0             0.0           0.0               0.0  ...   

      InternetService_DSL  InternetService_Fiber optic  InternetService_No  \
6290                  0.0                          0.0                 1.0   
520                   0.0                          1.0                 0.0   
5748                  0.0                          0.0                 1.0   
2315                  0.0                          1.0                 0.0   
1412                  0.0                          1.0                 0.0   

      Contract_Monthtomonth  Contract_One year  Contract_Two year  \
6290                    0.0                0.0                1.0   
520                     1.0                0.0                0.0   
5748                    0.0                0.0                1.0   
2315                    1.0                0.0                0.0   
1412                    1.0                0.0                0.0   

      PaymentMethod_Bank transfer (automatic)  \
6290                                      1.0   
520                                       0.0   
5748                                      0.0   
2315                                      0.0   
1412                                      0.0   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
6290                                    0.0                             0.0   
520                                     0.0                             1.0   
5748                                    0.0                             0.0   
2315                                    1.0                             0.0   
1412                                    0.0                             1.0   

      PaymentMethod_Mailed check  
6290                         0.0  
520                          0.0  
5748                         1.0  
2315                         0.0  
1412                         0.0  

[5 rows x 26 columns]

2.2 Logistic Regression

In [32]:
def log_reg(X_train, X_test, y_train, y_test, weights):
    if weights == -1:
        model = LogisticRegression()
    else:
        model = LogisticRegression(class_weight={0:weights[0], 1:weights[1]})
    
    model.fit(X_train, y_train)
    acc = model.score(X_test, y_test)
    print('Accuracy', acc, '\n')

    y_pred = model.predict(X_test)
    print('preds', y_pred[:5], '\n')

    cl_rep = classification_report(y_test, y_pred)
    print(cl_rep)

In [33]:
weights = -1
log_reg(X_train, X_test, y_train, y_test, weights)

ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

In [ ]:
_